examples/server/server.cpp

```diff
@@ -1,790 +1,923 @@
+    return i;
+}
+
+enum stop_type {
+    STOP_FULL,
+    STOP_PARTIAL,
+};
+
+bool ends_with(const std::string & str, const std::string & suffix) {
+    return str.size() >= suffix.size() &&
+        0 == str.compare(str.size() - suffix.size(), suffix.size(), suffix);
+}
+
+size_t find_partial_stop_string(const std::string & stop, const std::string & text) {
+    if (!text.empty() && !stop.empty()) {
+        const char text_last_char = text.back();
+        for (int64_t char_index = stop.size() - 1; char_index >= 0; char_index--) {
+            if (stop[char_index] == text_last_char) {
+                const std::string current_partial = stop.substr(0, char_index + 1);
+                if (ends_with(text, current_partial)) {
+                    return text.size() - char_index - 1;
+                }
+            }
         }
-        new_prompt_len ++;
-      }
     }
-    if(n_past > 0 && params.interactive) {
-      n_remain -= new_prompt_len;
-    }
-    if ((int)embd_inp.size() > params.n_ctx - 4)
-    {
-      return false;
+    return std::string::npos;
+}
+
+template<class Iter>
+static std::string tokens_to_str(llama_context * ctx, Iter begin, Iter end) {
+    std::string ret;
+    for (; begin != end; ++begin) {
+        ret += llama_token_to_str(ctx, *begin);
     }
-    has_next_token = true;
-    return true;
-  }
-
-  void beginCompletion()
-  {
-    if(n_remain == 0) {
-      // number of tokens to keep when resetting context
-      if (params.n_keep < 0 || params.n_keep > (int)embd_inp.size())
-      {
-        params.n_keep = (int)embd_inp.size();
-      }
+    return ret;
+}
+
+static void server_log(const char * level, const char * function, int line, const char * message, nlohmann::ordered_json extra) {
+    nlohmann::ordered_json log {
+        { "timestamp", time(NULL) },
+        { "level", level },
+        { "function", function },
+        { "line", line },
+        { "message", message },
+    };
+
+    if (!extra.empty()) {
+        log.merge_patch(extra);
     }
-    n_remain = params.n_predict;
-  }
-
-  llama_token nextToken() {
-    llama_token result = -1;
-    if (embd.size() > 0)
-    {
-      if (n_past + (int)embd.size() > params.n_ctx)
-      {
-        // Reset context
-        const int n_left = n_past - params.n_keep;
-        n_past = std::max(1, params.n_keep);
-        processed_tokens.erase(processed_tokens.begin() + n_past, processed_tokens.end());
-        embd.insert(embd.begin(), last_n_tokens.begin() + params.n_ctx - n_left / 2 - embd.size(), last_n_tokens.end() - embd.size());
-      }
-      for (int i = 0; i < (int)embd.size(); i += params.n_batch)
-      {
-        int n_eval = (int)embd.size() - i;
-        if (n_eval > params.n_batch)
-        {
-          n_eval = params.n_batch;
+
+    std::string str = log.dump(-1, ' ', false, json::error_handler_t::replace);
+    fprintf(stdout, "%.*s\n", (int)str.size(), str.data());
```

Comment:
```suggestion
    fprintf(stdout, "%.*s\n", (int)str.size(), str.data());
    fflush(stdout);
```

Reply:
nitpick: why are we doing the printf like that? Is it somehow faster/used like that elsewhere? and if all we do is print a string (+nl), why printf in the first place?

Reply:
`printf` is used in other places as well in llama.cpp, I don't think `std::cout` is used. It could also be done with a `std.c_str()` but that may cause an allocation and it's not strictly needed. We could use `fwrite` maybe then?

Or what do you propose?